File: simulation/script/miner_coin.py
import hashlib
import json
import time
from math import radians, sin, cos, sqrt, atan2

# Full time run simulation 
# begin_time = float(input("Enter beginning time parameter in running command: "))
# end_time = float(input("Enter ending time parameter in running command: "))
begin_time  = 0
end_time = 3600

# Define driver class
class Driver:
    def __init__(self, id, distance, time, coin):
        self.id = id
        self.distance = float(distance)
        self.time = float(time)
        self.coin = int(coin)

    def display(self):
        print(f"id: {self.id}, Distance: {self.distance}, Time: {self.time}, Coin: {self.coin}")

# Define vehicle class
class Vehicle:
    def __init__(self, vehicle, time):
        self.id = vehicle["id"]
        self.x = vehicle["x"]
        self.y = vehicle["y"]
        self.angle = vehicle["angle"]
        self.type = vehicle["type"]
        self.speed = vehicle["speed"]
        self.pos = vehicle["pos"]
        self.lane = vehicle["lane"]
        self.slope = vehicle["slope"]
        self.time = float(time)

# Hash string by sha512
def sha512(input_string):
    return hashlib.sha512(input_string.encode()).hexdigest()

# Get data from json and return a list of vehicle in a period of time
def get_data_from_json(begin, end):
    data = data_json["fcd-export"]["timestep"]

    data_list = []

    for element in data:
        time = float(element["time"])
        if begin <= time <= end:
            # Having an object
            if not isinstance(element["vehicle"], list):
                # Push data to the list
                data_list.append(Vehicle(element["vehicle"], element["time"]))
            # Having an object list
            else:
                # Push data to the list
                for v in element["vehicle"]:
                    data_list.append(Vehicle(v, element["time"]))

    return data_list

# Classify data of a node, return a new array that is classified
def classify_list(drivers):
    new_drivers = []
    check = [False] * len(drivers)

    for i in range(len(drivers)):
        lst = []
        if not check[i]:
            lst.append(drivers[i])
            check[i] = True
            for j in range(i + 1, len(drivers)):
                if drivers[i].id == drivers[j].id and not check[j]:
                    lst.append(drivers[j])
                    check[j] = True
            new_drivers.extend(lst)

    return new_drivers

# Calculate distance by haversine formula (meters)
def haversine(lat1, lon1, lat2, lon2):
    # Bán kính của Trái Đất (đơn vị: km)
    r = 6371.0
    lat1 = float(lat1)
    lon1 = float(lon1)
    lat2 = float(lat2)
    lon2 = float(lon2)

    # Chuyển đổi độ sang radian
    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    # Sự chênh lệch giữa vĩ độ và kinh độ
    dlon = lon2 - lon1
    dlat = lat2 - lat1

    # Sử dụng công thức Haversine
    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))
    distance = r * c * 1000  # The distance in meters

    return distance

# Calculate distance of a vehicle list, return a driver list
def calculate_distance_list(vehicles, distance, end):
    drivers = []
    d = 0
    c = 0
    for idx in range(1, len(vehicles)):
        if vehicles[idx].id == vehicles[idx - 1].id:
            timestep = vehicles[idx].time - vehicles[idx - 1].time
            round_time = round(timestep, 1)

            if round_time == 0.1:
                d += haversine(vehicles[idx].x, vehicles[idx].y, vehicles[idx - 1].x, vehicles[idx - 1].y)
                if d >= distance:
                    c += int(d / distance)
                    d %= distance

        if idx < len(vehicles) - 1 and vehicles[idx - 1].id != vehicles[idx].id:
            dr = Driver(vehicles[idx - 1].id, d, end, c)
            drivers.append(dr)
            d = 0
            c = 0
        elif idx == len(vehicles) - 1:
            dr = Driver(vehicles[idx - 1].id, d, end, c)
            drivers.append(dr)
            d = 0
            c = 0

    return drivers

# Return satisfying node proof of driving
def rule(drivers):
    node_pod = []

    w = sum(d.coin for d in drivers) / len(drivers)
    hash_w = sha512(str(w))

    for driver in drivers:
        if driver.coin != 0:
            hash_current = sha512(str(driver.coin))
            if hash_current <= hash_w:
                node_pod.append(driver)

    return node_pod

def count_number_of_vehicle(list_vehicle):
    count = 0
    if list_vehicle:
        count = 1
        for i in range(len(list_vehicle) - 1):
            if list_vehicle[i].id != list_vehicle[i + 1].id:
                count += 1
    return count

def main():
    begin = begin_time
    end = end_time
    timeslot = end - begin
    distance = 2000
    count = int(end_time / timeslot)
    output = []

    input_data = get_data_from_json(begin, end)
    class_list = classify_list(input_data)
    distance_list = calculate_distance_list(class_list, distance, end)
    n_pod = rule(distance_list)

    print(f"\nTime begin = {begin}, Time end = {end}")
    print(f"Number of node POD = {len(n_pod)}")

    # Statistics
    all_data_json = get_data_from_json(0, end_time)
    classify_all_data = classify_list(all_data_json)
    number_of_vehicle = count_number_of_vehicle(classify_all_data)
    print(f"Number of vehicle: {number_of_vehicle}")

if __name__ == "__main__":
    begin = time.time()
    with open("../data/vehicle1.json") as json_file:
        data_json = json.load(json_file)
    main()
    end = time.time()
    print("The time of execution of above program is :",
      (end-begin) * 10**3, "ms")

File: simulation/script/MinerCoin.js
const crypto = require("crypto");
const maxTime = require("process");
const fs = require('fs');

// Full time run simulation 
const timeslot = parseFloat(maxTime.argv[2]);
const beginTime = parseFloat(maxTime.argv[3]);
const endTime = parseFloat(maxTime.argv[4]);
const distance = parseFloat(maxTime.argv[5]);
const totalTime = parseFloat(maxTime.argv[6]);
const numVehicles = parseFloat(maxTime.argv[7]);

const filename = "../data/vehicle" + totalTime.toString() + "_" + numVehicles.toString() + ".json";
// const dataJson = require("../data/vehicle1.json");
const dataJson = require(filename);
// Define driver class
class Driver {
    constructor(id, distance, time, coin) {
        this.id = id;
        this.distance = Number(distance);
        this.time = Number(time);
        this.coin = Number(coin);
    }

    display() {
        console.log(
            "id: " +
                this.id +
                ", Distance: " +
                this.distance +
                ", Time: " +
                this.time +
                " , Coin: " +
                this.coin
        );
    }
}

// Define vehicle class
class Vehicle {
    constructor(vehicle, time) {
        this.id = vehicle.id;
        this.x = vehicle.x;
        this.y = vehicle.y;
        this.angle = vehicle.angle;
        this.type = vehicle.type;
        this.speed = vehicle.speed;
        this.pos = vehicle.pos;
        this.lane = vehicle.lane;
        this.slope = vehicle.slope;
        this.time = Number (time);
    }
}

// Hash string by sha512
function sha512(inputString) {
    return crypto.createHash("sha512").update(inputString).digest("hex");
}

function getFistElement(drivers) {
    return drivers.map((subarray) => subarray[0]);
}

// Get data from json and return list of vehicle in a period of time
function getDataFromJson(begin, end) {
    const data = dataJson["fcd-export"]["timestep"];

    dataList = [];

    data.forEach((element) => {
        time = Number(element.time);
        if (time >= begin && time <= end) {
            // Having a object
            if (element.vehicle.length == undefined) {
                // Push data to list
                dataList.push(new Vehicle(element.vehicle, element.time));
            }
            // Having object list
            else {
                // Push data to list
                element.vehicle.forEach((v) => {
                    dataList.push(new Vehicle(v, element.time));
                });
            }
        }
    });

    return dataList;
}

function getAllDataJson() {
    const data = dataJson["fcd-export"]["timestep"];

    dataList = [];

    data.forEach((element) => {
        // Having a object
        if (element.vehicle.length === undefined) {
            // Push data to list
            dataList.push(new Vehicle(element.vehicle, element.time));
        }
        // Having object list
        else {
            // Push data to list
            element.vehicle.forEach((v) => {
                dataList.push(new Vehicle(v, element.time));
            });
        }
    });

    return dataList;
}

// Classify data of a node, return new array is classified
function classifyList(drivers) {
    newDrivers = [];

    check = [];

    for (let i = 0; i < drivers.length; i++) {
        check.push(false);
    }

    for (let i = 0; i < drivers.length; i++) {
        list = [];
        if (check[i] == false) {
            list.push(drivers[i]);
            check[i] = true;
            for (let j = i + 1; j < drivers.length; j++) {
                if (drivers[i].id === drivers[j].id && check[j] == false) {
                    list.push(drivers[j]);
                    check[j] = true;
                }
            }
            list.forEach((l) => newDrivers.push(l));
        }
    }

    // newDrivers.forEach((v) => {console.log(v)})
    // console.log("DONE Classify list: Length = " + newDrivers.length)
    return newDrivers;
}

// calculate distance by haversine formal (miles)
function haversine(lat1, lon1, lat2, lon2) {
    const R = 6371; // Radius of the Earth in kilometers
    const dLat = (lat2 - lat1) * (Math.PI / 180);
    const dLon = (lon2 - lon1) * (Math.PI / 180);

    const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
        Math.cos(lat1 * (Math.PI / 180)) * Math.cos(lat2 * (Math.PI / 180)) *
        Math.sin(dLon / 2) * Math.sin(dLon / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

    var distance = R * c; // The distance in kilometers
    distance *= 1000; // km => m
    return distance;
}

// Calculate distance of a vehicle list, return a driver list
function calculateDistanceList(vehicles, distance, end) {
    // console.log("Vehicle length in calculate distance: " + vehicles.length);
    // vehicles.forEach((v) => {console.log(v)})
    // console.log("INPUT calculate distance: Length = " + vehicles.length);
    
    var drivers = [];
    d = 0;
    c = 0;
    for (let idx = 1; idx < vehicles.length; idx++) {
        // console.log("calculateDistanceList: ")
        if (vehicles[idx].id === vehicles[idx - 1].id) {
            timestep = vehicles[idx].time - vehicles[idx - 1].time;
            roundTime = parseFloat(timestep.toFixed(1));

            if (roundTime === 0.1) {
                d += haversine(vehicles[idx].x, vehicles[idx].y, vehicles[idx-1].x, vehicles[idx-1].y);
                // console.log("Distance = ", d)
                if (d >= distance) {
                    c = c + parseInt(d / distance);
                    d = d % distance;
                }
                // console.log("Coin = ", c)
            }
        } 

        if (idx < vehicles.length - 1) {
            if (vehicles[idx - 1].id != vehicles[idx].id) {
                const dr = new Driver(vehicles[idx - 1].id, d, end, c);
                // console.log(dr)
                drivers.push(dr);
                // console.log("Dr Coin", dr.coin)
                d = 0;
                c = 0;
            }  
        } else if (idx == vehicles.length - 1) {
            const dr = new Driver(vehicles[idx - 1].id, d, end, c);
            // console.log(dr)
            drivers.push(dr);
            // console.log("Dr Coin", dr.coin)
            d = 0;
            c = 0;
        }
           
    }

    // drivers.forEach(v => console.log(v))
    console.log("DONE calculate distance: Number of vehicle = " + drivers.length);

    return drivers;
}

// Return satisfy node proof of driving
function rule(drivers) {
    // console.log("INPUT rule: number of drivers = " + drivers.length);

    nodePod = [];

    w = 0;
    drivers.forEach(d => {
        // console.log("d: " + d.coin)
        w += d.coin;
    })
    // console.log("w: " + w)
    // console.log("driver.length:  " + drivers.length)
    w = w / drivers.length;

    //  Get hash value of w
    let hashW = sha512(w.toString());

    for (let i = 0; i < drivers.length; i++) {
        //  Get hash value of driver
        if (drivers[i].coin != 0) {
            hashCurrent = sha512(drivers[i].coin.toString());

            if (hashCurrent.localeCompare(hashW) <= 0) {
                nodePod.push(drivers[i]);
            }
        }
    }

    console.log("DONE rule: Number of node POD = " + nodePod.length);
    // nodePod.forEach((v) => console.log(v));
    return nodePod;
}

function countNumberOfVehicle(listVehicle) {
    // Starting counter at first element in list vehicle
    let count = 0;
    if (listVehicle.length != 0) {
        count = 1;
        for (let i = 0; i < listVehicle.length - 1; i++) 
            if (listVehicle[i].id === listVehicle[i+1].id) continue
            else count++
    }
        
    return count;
}

function main() {
    let begin = beginTime;                  // Beginning time calculates distances and coins in 1 round (s)
    let end = endTime;     
    if (isNaN(end)) console.log("Warning: Please enter beginning time parameter in running command");
    if (isNaN(begin)) console.log("Warning: Please enter ending time parameter in running command");
    if (isNaN(distance)) console.log("Warning: Please enter distance parameter in running command");
    if (endTime)
    console.log("\nTime begin = " + begin + " Time end = " + end);

    let d = distance;            // Distance to earning coin (m)
    const count = Math.ceil(endTime / timeslot);
    // console.log("Count: " + count);

    const output = [];

    const inputData = getDataFromJson(begin, end);

    const classList = classifyList(inputData);
    
    const distanceList = calculateDistanceList(classList, distance, end);

    

    const nPOD = rule(distanceList);
    // console.log(inputData)
    // console.log(classList.length);
    

    // for (let i = 0; i < count; i++) {
    //     let coins = [];

    //     const inputData = getDataFromJson(begin, end);

    //     const classList = classifyList(inputData);
        
    //     const distanceList = calculateDistanceList(classList, distance, end);

    //     const nPOD = rule(distanceList);

    //     nPOD.forEach((v) => {
    //         // console.log(v)
    //         output.push(v);
    //     });

    //     if (end === 3600) {
    //         console.log("Time end = " + end);
    //         console.log(inputData)
    //         // console.log(classList)
    //         // console.log(distanceList)
    //         // console.log(nPOD)
    //     }

    //     begin += timeslot;
    //     end += timeslot;
    //     if (end > endTime) end = endTime;
    // }

    // output.forEach((v) => console.log(v));

    // Statistics
    // var allDataJson = getAllDataJson();
    // var classifyAllData = classifyList(allDataJson);
    // var numberOfVehicle = countNumberOfVehicle(classifyAllData);
    // console.log("Number of vehicle ", numberOfVehicle)
    
    
    // Statistic
    const header = ['Timeslot', 'Begin', 'End', 'Distance', 'Total node', 'Node received coin'];
    const dataArrays = [
    [timeslot, begin, end, distance, distanceList.length, nPOD.length]]
    

    var stream = fs.createWriteStream("../data/data_statistic.csv", {'flags': 'a'});
    
    // Running once time to add header
    // stream.once('open', function(fd) {
    //     stream.write(header+"\r\n");
    //   });
    
    stream.once('open', function(fd) {
      stream.write(dataArrays+"\r\n");
    });
}
const start = Date.now();
main();
const end = Date.now();
console.log(`Execution time: ${end - start} ms`);

module.exports = {
    rule
};


